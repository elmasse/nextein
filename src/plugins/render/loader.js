const loaderUtils = require('loader-utils')
const { resolve } = require('path')
/**
 * loads a generated file that allows us to require all render plugin in react.
 *
 * results in a file that looks like:
 *
 * ```js
 * module.exports = {
 *   "some-plugin": require('some-plugin').render,
 *   "some-other-plugin": require('some-other-plugin').render
 * }
 * ```
 */
module.exports = function () {
  const { plugins = [] } = loaderUtils.getOptions(this)

  return `module.exports = {
    "__PLUGINS__": ${JSON.stringify(plugins)},
    ${plugins
      .filter(p => p.renderer)
      .map(({ resolved, name }) => `"${name}": require(${JSON.stringify(resolve(resolved, 'render'))}).render`)
      .join(',\n')
    }
  }`
}
