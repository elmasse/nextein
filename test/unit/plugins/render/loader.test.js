jest.mock('loader-utils')
jest.mock('path')

import loaderUtils from 'loader-utils'
import { resolve } from 'path'

//SUT
import loader from '../../../../src/plugins/render/loader'


describe('render webpack loader', () => {
  
  const pluginA = { name: 'pluginA', resolved: '/test/pluginA'  }
  const withRenderer = { name: 'pluginB', resolved: '/test/pluginB', renderer: true  }

  beforeEach(() => {
    resolve.mockImplementation((...args) => {
      const original = jest.requireActual('path').resolve
      return original(...args)
    })  
  })

  afterEach(() => {
    resolve.mockRestore()
  })

  test('export with no plugins', () => {
    loaderUtils.getOptions.mockReturnValue({ plugins: [] })
    
    const result = loader()
    
    expect(result).toMatchSnapshot()
  })

  test('export plugins', () => {
    loaderUtils.getOptions.mockReturnValue({ plugins: [pluginA] })
    
    const result = loader()
    
    expect(result).toMatchSnapshot()
    expect(resolve).not.toBeCalled()
  })

  test('export plugins with renderer ', () => {
    loaderUtils.getOptions.mockReturnValue({ plugins: [pluginA, withRenderer] })

    const result = loader()
    
    expect(result).toMatchSnapshot()
    expect(resolve).toBeCalledWith(withRenderer.resolved, 'render')

  })
})

describe('render webpack loader Win32 FS', () => {
  const pluginA = { name: 'pluginA', resolved: 'C:\\test\\pluginA'  }
  const withRenderer = { name: 'pluginB', resolved: 'C:\\test\\pluginB', renderer: true  }

  beforeEach(() => {
    resolve.mockImplementation((...args) => {
      const original = jest.requireActual('path').win32.resolve
      return original(...args)
    })
  })

  afterEach(() => {
    resolve.mockRestore()
  })

  test('export with no plugins', () => {
    loaderUtils.getOptions.mockReturnValue({ plugins: [] })
    
    const result = loader()
    
    expect(result).toMatchSnapshot()
  })

  test('export plugins', () => {
    loaderUtils.getOptions.mockReturnValue({ plugins: [pluginA] })
    
    const result = loader()
    
    expect(result).toMatchSnapshot()
    expect(resolve).not.toBeCalled()
  })

  test('export plugins with renderer ', () => {
    loaderUtils.getOptions.mockReturnValue({ plugins: [pluginA, withRenderer] })

    const result = loader()

    expect(result).toMatchSnapshot()
    expect(resolve).toBeCalledWith(withRenderer.resolved, 'render')

  })
})
