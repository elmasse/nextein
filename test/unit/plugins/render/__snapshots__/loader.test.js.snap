// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render webpack loader Win32 FS export plugins 1`] = `
"module.exports = {
    \\"__PLUGINS__\\": [{\\"name\\":\\"pluginA\\",\\"resolved\\":\\"C:\\\\\\\\test\\\\\\\\pluginA\\"}],
    
  }"
`;

exports[`render webpack loader Win32 FS export plugins with renderer  1`] = `
"module.exports = {
    \\"__PLUGINS__\\": [{\\"name\\":\\"pluginA\\",\\"resolved\\":\\"C:\\\\\\\\test\\\\\\\\pluginA\\"},{\\"name\\":\\"pluginB\\",\\"resolved\\":\\"C:\\\\\\\\test\\\\\\\\pluginB\\",\\"renderer\\":true}],
    \\"pluginB\\": require(\\"C:\\\\\\\\test\\\\\\\\pluginB\\\\\\\\render\\").render
  }"
`;

exports[`render webpack loader Win32 FS export with no plugins 1`] = `
"module.exports = {
    \\"__PLUGINS__\\": [],
    
  }"
`;

exports[`render webpack loader export plugins 1`] = `
"module.exports = {
    \\"__PLUGINS__\\": [{\\"name\\":\\"pluginA\\",\\"resolved\\":\\"/test/pluginA\\"}],
    
  }"
`;

exports[`render webpack loader export plugins with renderer  1`] = `
"module.exports = {
    \\"__PLUGINS__\\": [{\\"name\\":\\"pluginA\\",\\"resolved\\":\\"/test/pluginA\\"},{\\"name\\":\\"pluginB\\",\\"resolved\\":\\"/test/pluginB\\",\\"renderer\\":true}],
    \\"pluginB\\": require(\\"/test/pluginB/render\\").render
  }"
`;

exports[`render webpack loader export with no plugins 1`] = `
"module.exports = {
    \\"__PLUGINS__\\": [],
    
  }"
`;
